kv/etcd/etcd.go

```diff
@@ -0,0 +1,220 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package etcd
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/golang/protobuf/proto"
+	"github.com/m3db/m3cluster/kv"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/retry"
+	"golang.org/x/net/context"
+)
+
+var noopCancel func()
+
+// NewStore creates a kv store based on etcd
+func NewStore(opts Options) (kv.Store, error) {
+	etcd, err := clientv3.New(clientv3.Config{
```

Comment:
Might it be worth having callers pass in the `etcd` client so they can configure it however they may like?  e.g. `NewStore(client *clientv3.Client, opts Options) (kv.Store, error)`

Just in case in there is something they would prefer to initialize themselves and/or pass in a shared client they already have?

Also: we could go one step further and define the interface that matches methods of `*clientv3.Client` so that technically you could pass a mock etcd client in here, especially for testing purposes this would be most ideal?

We could define the following interface that matches the `*clientv3.Client` methods so it can be seamlessly passed in:

```go
type EtcdClient interface {
  clientv3.KV
  clientv3.Watcher

  Close() error
  Ctx() context.Context
  Endpoints() (eps []string)
  SetEndpoints(eps ...string)
  Sync(ctx context.Context) error
  Dial(endpoint string) (*grpc.ClientConn, error)
  ActiveConnection() *grpc.ClientConn
}
```